src/util.rs

```diff
@@ -313,6 +297,44 @@ pub trait ReadExt<'data> {
     fn read_u64(&mut self) -> Option<u64> {
         self.read_pod::<u64>()
     }
+
+    /// Read a `u128` encoded as unsigned variable length little endian base 128
+    /// value.
+    ///
+    /// The function returns the value read along with the number of bytes
+    /// consumed.
+    fn read_u128_leb128(&mut self) -> Option<(u128, u8)> {
+        let mut shift = 0;
+        let mut value = 0u128;
+        while let Some(bytes) = self.read_slice(1) {
+            if let [byte] = bytes {
+                value |= ((byte & 0b0111_1111) as u128) << shift;
+                shift += 7;
+                if (byte & 0b1000_0000) == 0 {
+                    return Some((value, shift / 7));
+                }
+            } else {
+                unreachable!()
+            }
+        }
+        None
+    }
+
+    /// Read a `u128` encoded as signed variable length little endian base 128
+    /// value.
+    ///
+    /// The function returns the value read along with the number of bytes
+    /// consumed.
+    fn read_i128_leb128(&mut self) -> Option<(i128, u8)> {
+        let (value, shift) = self.read_u128_leb128()?;
+        let sign_mask = 1u128 << (shift * 7 - 1);
+        if (value & sign_mask) != 0 {
+            // negative
+            Some(((value as i128) - ((sign_mask << 1) as i128), shift))
+        } else {
+            Some((value as i128, shift))
+        }
```

Comment:
so in C I'd write this more elegantly (I think Rust should allow something same with unchecked shifts):

```
let sign_bits = 128 - shift * 7;
let value = ((value as i128) << sign_bits) >> sign_bits;
```

Reply:
Adopted.